.github/workflows/pull_request.yml

```diff
@@ -1,51 +1,135 @@
         uses: actions/checkout@v4
+
       - name: Check if container definition has changed
         id: changed-container-def
-        uses: tj-actions/changed-files@v41
+        uses: tj-actions/changed-files@48d8f15b2aaa3d255ca5af3eba4870f807ce6b3c # v43.0.2
         with:
           files_yaml: |
             containerdef:
               - container/container.def
+
       - name: Build container if definition has changed
         if: steps.changed-container-def.outputs.containerdef_any_changed == 'true'
         run: |
             echo ${{ needs.generate_matrix.outputs.matrix }}
             sudo -E singularity build container/base.sif container/container.def
+
+      - name: Upload base container
+        if: steps.changed-container-def.outputs.containerdef_any_changed == 'true'
+        uses: actions/upload-artifact@v4
+        with:
+            name: base-container
+            path: container/base.sif
+
+  setup:
+    runs-on: ubuntu-latest
+    needs: [ generate_matrix, build_base_image ]
+    if: ${{ needs.generate_matrix.outputs.matrix != '{"include":[]}' }}
+    environment: Gadi
+    steps:
+      - name: Check Gadi Github Environment configuration
+        run: |
+          vars_unset=false
+          if [ -z "${{ vars.ADMIN_DIR }}" ]; then
+            echo "::error::vars.ADMIN_DIR is unset. ADMIN_DIR is the base directory which will contain the staging and logs sub-directories"
+            vars_unset=true
+          fi
+          if [ -z "${{ vars.CONDA_BASE }}" ]; then
+            echo "::error::vars.CONDA_BASE is unset. CONDA_BASE is the base deployment directory which will contain apps and module sub-directories"
+            vars_unset=true
+          fi
+          if [ -z "${{ vars.PROJECT }}" ]; then
+            echo "::error::vars.PROJECT is unset. PROJECT is the project code passed to build+test qsub calls"
+            vars_unset=true
+          fi
+          if [ -z "${{ vars.STORAGE }}" ]; then
+            echo "::error::vars.STORAGE is unset. STORAGE is the storage flags passed to build+test qsub calls"
+            vars_unset=true
+          fi
+          if [ -z "${{ vars.APPS_USERS_GROUP }}" ]; then
+            echo "::error::vars.APPS_USERS_GROUP is unset. APPS_USERS_GROUP is used for setting file permissions"
+            vars_unset=true
+          fi
+          if [ -z "${{ vars.APPS_OWNERS_GROUP }}" ]; then
+            echo "::error::APPS_OWNERS_GROUP is unset. APPS_OWNERS_GROUP is used for setting file permissions"
+            vars_unset=true
+          fi
+          if [ -z "${{ secrets.REPO_PATH }}" ]; then
+            echo "::error::secrets.REPO_PATH is unset. REPO_PATH is the path where this repository is synced to"
+            vars_unset=true
+          fi
+
+          if [ "$vars_unset" == "true" ]; then
+            echo "::error::Required vars in Gadi Github Environment are unset."
+            exit 1
+          fi
+
+      - name: Checkout repository
+        uses: actions/checkout@v4
```

Comment:
One of the things that I'm not too sure about when using the relative `uses: ./.github/workflows/thingo.yml` as opposed to the version-specific `uses: access-nri/model-release-condaenv/.github/workflows/thingo.yml@main` is that in the case where it is `on.pull_request` - the checkout that you do here means that the version of [the deploy workflow that you use here](https://github.com/ACCESS-NRI/model-release-condaenv/pull/1/files#diff-a0fe23534b616d51ce686d2a1bcd1a78bc75074aef1a2f6ee96c9469991e1a4cR134-R136) would be the pull request version, rather than the `@main` version, for example.

Reply:
I think the relative paths were useful to the use the test the changes in this PR. But maybe if the `build` and `test` workflows work,  the workflow changes should be separated into a different PR and merged first, with the versioned values `@main`?

Reply:
Yeah that will work :)